opensearch_py_ml/ml_models/metrics_correlation/dip.py

```diff
@@ -0,0 +1,407 @@
+                        ig = ix + 1
+                        ih = iv - 1
+                else:
+                    lcmiv1 = lcm[iv - 1]
+                    dx = (x[gcmix] - x[lcmiv1]) * (lcmiv - lcmiv1) / (
+                        x[lcmiv] - x[lcmiv1]
+                    ) - (gcmix - lcmiv1 - 1)
+                    ix -= 1
+                    if dx >= d:
+                        d = dx
+                        ig = ix + 1
+                        ih = iv
+
+                if ix < 0:
+                    ix = 0
+                if iv > l_lcm:
+                    iv = l_lcm
+
+                if gcm[ix] == lcm[iv]:
+                    break
+        else:
+            d = 1.0
+
+        if d < dstar:
+            break
+
+        # compute dip for cvx minorant
+        dip_l = 0.0
+        for j in range(ig, l_gcm):
+            max_t = 1.0
+            jb = gcm[j + 1]
+            je = gcm[j]
+            if je - jb > 1 and x[je] != x[jb]:
+                C = torch.log(torch.tensor(je - jb)) - torch.log(
+                    x[je] - x[jb]
+                )  # log space for numeric stability
+                for jjj in range(jb, je + 1):
+                    xx = torch.log(x[jjj] - x[jb])
+                    t = (jjj - jb + 1) - torch.exp(xx + C)
+                    if max_t < t:
+                        max_t = t
+
+            if dip_l < max_t:
+                dip_l = max_t
+
+        # compute dip for cnc majorant
+        dip_u = 0.0
+        for j in range(ih, l_lcm):
+            max_t = 1.0
+            jb = lcm[j]
+            je = lcm[j + 1]
+            if je - jb > 1 and x[je] != x[jb]:
+                C = torch.log(torch.tensor(je - jb)) - torch.log(
+                    x[je] - x[jb]
+                )  # log space for numeric stability
+                for jjj in range(jb, je + 1):
+                    xx = torch.log(x[jjj] - x[jb])
+                    t = torch.exp(xx + C) - (jjj - jb - 1)
+                    if max_t < t:
+                        max_t = t
+
+            if dip_u < max_t:
+                dip_u = max_t
+
+        # print('dip_l, dip_u = ({},{})'.format(dip_l, dip_u))
+
+        # update current max dip
+        if dip_u > dip_l:
+            dipnew = dip_u
+        else:
+            dipnew = dip_l
+
+        if dstar < dipnew:
+            dstar = dipnew
+
+        if low == gcm[ig] and high == lcm[ih]:
+            # print('no improvement in low / high -> end')
+            break
+        else:
+            low = gcm[ig]
+            high = lcm[ih]
+
+    # end logic
+    dstar = dstar / (2 * n)
+
+    return dstar  # mn, mj, gcm, lcm, dstar
+
+
+def diptest(x: torch.Tensor) -> Tuple[float, float]:
+    """
+    Testing the dip
+
+    :param x:
+    :type x: torch.Tensor
+    :return:
+    :rtype: Tuple[float, float]
+    """
+
+    # torchScript doesn't support tensor values as global variable so we can't put these
+    # constant values in utils.py
+
+    N_VALS: torch.Tensor = torch.tensor(
```

Comment:
The p-value lookup table here has a sample size (i.e. length of time series) range of $T=100$ to $T=50000$. The upper bound is probably fine but we should consider whether the lower bound is too high. The method doesn't break if the input has a length outside of these bounds, but the p-value calculation and thus event selection can be inaccurate. We should consider whether a wider range needs to be supported.